The Basics

NumPy’s main object is the homogeneous multidimensional array. It is a table of elements (usually numbers), all of the same type, indexed by a tuple of positive integers. In NumPy dimensions are called axes.
For example, the coordinates of a point in 3D space [1, 2, 1] has one axis. That axis has 3 elements in it, so we say it has a length of 3. In the example pictured below, the array has 2 axes. The first axis has a length of 2, the second axis has a length of 3.
[[ 1., 0., 0.],
 [ 0., 1., 2.]]
NumPy’s array class is called ndarray. It is also known by the alias array. Note that numpy.array is not the same as the Standard Python Library class array.array, which only handles one-dimensional arrays and offers less functionality. The more important attributes of an ndarray object are:

ndarray.ndim
the number of axes (dimensions) of the array.

ndarray.shape
the dimensions of the array. This is a tuple of integers indicating the size of the array in each dimension. For a matrix with n rows and m columns, shape will be (n,m). The length of the shape tuple is therefore the number of axes, ndim.

ndarray.size
the total number of elements of the array. This is equal to the product of the elements of shape.

ndarray.dtype
an object describing the type of the elements in the array. One can create or specify dtype’s using standard Python types. Additionally NumPy provides types of its own. numpy.int32, numpy.int16, and numpy.float64 are some examples.

ndarray.itemsize
the size in bytes of each element of the array. For example, an array of elements of type float64 has itemsize 8 (=64/8), while one of type complex32 has itemsize 4 (=32/8). It is equivalent to ndarray.dtype.itemsize.

ndarray.data
the buffer containing the actual elements of the array. Normally, we won’t need to use this attribute because we will access the elements in an array using indexing facilities.

In [4]:
import numpy as np
a = np.arange(15).reshape(3, 5)

In [5]:
a

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14]])

In [7]:
dimension = a.ndim
shape = a.shape
size = a.size
data_type = a.dtype
form = a.itemsize
buffer = a.data

In [11]:
dimension

2

In [10]:
shape

(3, 5)

In [12]:
size

15

In [13]:
data_type

dtype('int32')

In [14]:
form

4

In [15]:
buffer

In [16]:
type(a)

numpy.ndarray

In [19]:
b = np.array([3.66, 7.0, 1])
b

array([3.66, 7.  , 1.  ])

In [20]:
c = np.array([(2, 3, 4), (4, 5, 6)])
c

array([[2, 3, 4],
       [4, 5, 6]])

In [25]:
d = np.array([[1, 2], (2, 3)], dtype = complex)
d

array([[1.+0.j, 2.+0.j],
       [2.+0.j, 3.+0.j]])

In [28]:
e = np.zeros([2, 4])
e

array([[0., 0., 0., 0.],
       [0., 0., 0., 0.]])

In [29]:
f = np.ones([2, 2, 2])
f

array([[[1., 1.],
        [1., 1.]],

       [[1., 1.],
        [1., 1.]]])

In [36]:
g = np.empty([5, 5, 1])
g

array([[[1.09421637e-311],
        [7.36157812e-322],
        [0.00000000e+000],
        [0.00000000e+000],
        [1.33511018e-306]],

       [[1.15998412e-028],
        [2.44171989e+232],
        [8.00801729e+159],
        [8.16124052e-153],
        [6.03104724e-154]],

       [[6.03105862e-154],
        [6.01391519e-154],
        [9.89069227e-096],
        [1.08266064e-095],
        [1.19683021e+141]],

       [[6.01347002e-154],
        [4.57668654e-072],
        [4.57669517e-072],
        [1.95361036e-109],
        [2.86752281e+161]],

       [[2.78225500e+296],
        [9.80058441e+252],
        [1.23971686e+224],
        [2.22524390e-306],
        [0.00000000e+000]]])

In [41]:
h = np.arange(10, 31, 5)
h

array([10, 15, 20, 25, 30])

In [43]:
i = np.linspace(3, 11.5, 17)
i

array([ 3.     ,  3.53125,  4.0625 ,  4.59375,  5.125  ,  5.65625,
        6.1875 ,  6.71875,  7.25   ,  7.78125,  8.3125 ,  8.84375,
        9.375  ,  9.90625, 10.4375 , 10.96875, 11.5    ])

In [47]:
x = i
j = np.sin(x)
print(j)

[ 0.14112001 -0.37987147 -0.79615098 -0.99297065 -0.91607692 -0.58666564
 -0.09553936  0.42192244  0.82308088  0.99735622  0.89670928  0.5488835
  0.04975741 -0.46308437 -0.84827645 -0.99964025 -0.87545217]


In [50]:
np.arange(10000)

array([   0,    1,    2, ..., 9997, 9998, 9999])

If an array is too large to be printed, NumPy automatically skips the central part of the array and only prints the corners:
>>>
>>> print(np.arange(10000))
[   0    1    2 ..., 9997 9998 9999]
>>>
>>> print(np.arange(10000).reshape(100,100))
[[   0    1    2 ...,   97   98   99]
 [ 100  101  102 ...,  197  198  199]
 [ 200  201  202 ...,  297  298  299]
 ...,
 [9700 9701 9702 ..., 9797 9798 9799]
 [9800 9801 9802 ..., 9897 9898 9899]
 [9900 9901 9902 ..., 9997 9998 9999]]
To disable this behaviour and force NumPy to print the entire array, you can change the printing options using set_printoptions.
>>>
>>> np.set_printoptions(threshold=np.nan)